### PR TITLE
KIP-392 Draft

### PR DESCRIPTION
**Pull request by mumrah**
Adds `ReplicaSelector` interface which is used by the broker to calculate the preferred read replica for consumers. The implementation is selected by the `replica.selector.class` broker config. Only the leader can calculate the preferred replica, and it only does so for regular consumer fetch requests. Three implementations are included here, but for the first phase we will probably only include the leader-only selector.

A new `client.rack` consumer configuration property is added which allows the consumer to indicate its rack.

FetchRequest now includes "rack_id" field which can optionally be specified by the consumer. FetchResponse includes an optional "preferred_read_replica" field for each partition in the response.

ReplicaManager was also modified to allow fetch requests from regular consumers to non-leader replicas if (and only if) the consumer includes "rack_id" field in the fetch request.

### CODE REVIEW

**Comment by mumrah** [comment]
retest this please

**Comment by mumrah** [comment]
retest this please

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 1]
```diff
@@ -1146,6 +1160,9 @@ private PartitionRecords parseCompletedFetch(CompletedFetch completedFetch) {
                 log.warn("Received unknown topic or partition error in fetch for partition {}", tp);
                 this.metadata.requestUpdate();
             } else if (error == Errors.OFFSET_OUT_OF_RANGE) {
+                // Go back to the leader
+                subscriptions.clearPreferredReadReplica(tp);
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:1164]: Should the client act on the out of range error if we are not fetching from the current leader? I am wondering if the logic should be something like this:
```
if (fetchingFromLeader)
 normalOutOfRangeValidation
else
 clearPreferredReadReplica
 validateCurrentOffset
```

**Comment by mumrah** [comment]
We're splitting this PR into two. First will be #6731 for the client side changes

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 2]
```diff
@@ -823,9 +831,24 @@ class ReplicaManager(val config: KafkaConfig,
                     fetchInfos: Seq[(TopicPartition, PartitionData)],
                     quota: ReplicaQuota = UnboundedQuota,
                     responseCallback: Seq[(TopicPartition, FetchPartitionData)] => Unit,
-                    isolationLevel: IsolationLevel) {
+                    isolationLevel: IsolationLevel,
+                    clientMetadata: ClientMetadata) {
     val isFromFollower = Request.isValidBrokerId(replicaId)
-    val fetchOnlyFromLeader = replicaId != Request.DebuggingConsumerId && replicaId != Request.FutureLocalReplicaId
+
+    val fetchOnlyFromLeader = {
```
Comment (mumrah) [inline core/src/main/scala/kafka/server/ReplicaManager.scala:838]: Let's just allow for fetching from followers in general, not dependent on rackId

**Pull request closed without merging**